[ADD] stock_picking_putaway_recompute

rousseldenis opened the pull request.

rousseldenis committed:
[ADD] stock_picking_putaway_recompute

rousseldenis committed:
[IMP] stock_picking_putaway_recompute: Don't authorize recompute with packages

As soon as there is a result package for the operation, the destination location
recomputation is not authorized as all operations contained in a package should go
to the same destination.

rousseldenis committed:
[IMP] stock_picking_putaway_recompute: Add a test with option deactivation

rousseldenis commented:
@lmignon @jbaudoux @phschmidt

rousseldenis committed:
[IMP] stock_picking_putaway_recompute: Apply first the move destination location

This is more correct as the original location before applying
the putaways is the stock move destination location.

simahawk commented on stock_picking_putaway_recompute/models/stock_move_line.py:
```diff
@@ -0,0 +1,58 @@
+    )
+
+    @api.depends(
+        "picking_type_id.allow_to_recompute_putaways", "picking_id.printed", "qty_done"
```
`allow_to_recompute_putaways` is not stored, should not stay here

grindtildeath replied:
Why? AFAIK you can use non stored computed fields as dependencies for other non stored computed fields :shrug: 

EDIT: BTW seems we need more dependencies, cf my comment below
EDIT 2: Actually it's a stored field on picking type :wink:

simahawk commented on stock_picking_putaway_recompute/tests/test_recompute_putaway.py:
```diff
@@ -0,0 +1,325 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo.fields import Command, first
+
+from odoo.addons.base.tests.common import BaseCommon
```
`BaseCommon` does too many things and it slows down setup. If you don't need company, currency, user, etc, use `TransactionCase` + `tracking_disable` or `DISABLED_MAIL_CONTEXT`.

simahawk reviewed (commented):
LG, minor remarks

grindtildeath commented on stock_picking_putaway_recompute/models/stock_move_line.py:
```diff
@@ -0,0 +1,58 @@
+            - have their qty_done field != 0
+        """
+        return self.filtered(
+            lambda line: line.picking_type_id.allow_to_recompute_putaways
```
Wouldn't it make more sense to have the filtering function defined at line level instead of one that filters out without the ability to override?

grindtildeath commented on stock_picking_putaway_recompute/models/stock_picking.py:
```diff
@@ -0,0 +1,38 @@
+        help="Technical field in order to display the Recompute Putaways button.",
+    )
+
+    def action_recompute_putaways(self) -> dict:
```
I don't know much about annotations but AFAIU this function must return a dict but doesn't return anything

grindtildeath commented on stock_picking_putaway_recompute/models/stock_picking.py:
```diff
@@ -0,0 +1,38 @@
+        Filter current recordset in order to get the pickings that can
+        """
+        return self.filtered(
+            lambda picking: picking.state == "assigned" and not picking.printed
```
Same as above about filtering function. Moreover, this does not check if recompute putaway is allowed at picking type level :confused: 

I feel like the filtering function from stock move should somehow call the filtering function from picking for consistency.

I would suggest to do as follows:
```
class StockMoveLine(models.Model):
    def _can_recompute_putaway(self):
        self.ensure_one()
        return self.picking_id._can_recompute_putaway() and not self.result_package_id and not self.qty_done

    def _filtered_can_recompute_putaways(self)
        return self.filtered(lambda li: li._can_recompute_putaway())


class StockPicking(models.Model):
    def _can_recompute_putaway(self):
        self.ensure_one()
        return self.picking_type_id.allow_to_recompute_putaways and self.state == "assigned" and not self.printed

    def _filtered_can_recompute_putaways(self)
        return self.filtered(lambda pick: pick._can_recompute_putaway())
```
What do you think?

grindtildeath committed:
[FIX] stock_picking_putaway_recompute: Harmonize computations and fix dependencies

Recomputation was allowed at picking level even if it was not allowed
on the picking type.
Make sure the computation on stock.move.line reuses the conditions defined
at picking level.

grindtildeath commented:
@rousseldenis I did the suggested changes here to speed up the merge of that PR: https://github.com/acsone/stock-logistics-workflow/pull/2

Can you please check and merge if that's good for you? :pray:

lmignon commented:
> @rousseldenis I did the suggested changes here to speed up the merge of that PR: [acsone#2](https://github.com/acsone/stock-logistics-workflow/pull/2)
> 
> Can you please check and merge if that's good for you? 🙏

done :smirk:

jbaudoux commented:
Module already merged in v18

/ocabot merge nobump
